lib/Optimizer/Transforms/PhaseFolding.cpp

```diff
@@ -0,0 +1,656 @@
+        return;
+
+      end_point = index;
+      pruneFrom(index);
+    }
+
+  public:
+    NetlistWrapper(Subcircuit *subcircuit, SmallVector<Operation *> *nl,
+                   Operation *anchor_point, Value def) {
+      this->nl = nl;
+      this->subcircuit = subcircuit;
+      this->def = def;
+      processFrom(getIndexOf(anchor_point));
+    }
+
+    void addNewAnchorPoint(Operation *op) {
+      auto index = getIndexOf(op);
+      if (index >= start_point)
+        return;
+      processFrom(index);
+    }
+
+    bool hasOps() { return end_point > start_point; }
+
+    void prune() { pruneFrom(end_point); }
+
+    Value getDef() { return def; }
+  };
+
+  Netlist *container;
+  SmallVector<NetlistWrapper *> qubits;
+  SetVector<Operation *> ops;
+  SmallVector<Operation *> ordered_ops;
+  Operation *start;
+  size_t num_rot_gates = 0;
+
+  void allocWrapper(Value ref, Operation *anchor_point) {
+    auto nlindex = container->getIndexOf(ref);
+    if (nlindex >= qubits.size())
+      for (auto i = qubits.size(); i < container->size(); i++)
+        qubits.push_back(nullptr);
+    auto *nlw = new NetlistWrapper(this, container->getNetlist(nlindex),
+                                   anchor_point, ref);
+    qubits[nlindex] = nlw;
+  }
+
+  /// @brief Gets the NetlistWrapper for ref, if it exists
+  /// @returns The NetlistWrapper for the Netlist for ref or
+  /// `nullptr` if no such Netlist exists
+  NetlistWrapper *getWrapper(Value ref) {
+    if (!isSupportedValue(ref))
+      return nullptr;
+
+    auto nlindex = container->getIndexOf(ref);
+    // Can still be nullptr if the wrapper hasn't been initialized
+    return qubits[nlindex];
+  }
+
+  void processNextAnchorPoint() {
+    auto next = anchor_points.back();
+    anchor_points.pop_back();
+    auto nl = getWrapper(next.first);
+    if (nl)
+      nl->addNewAnchorPoint(next.second);
+    else
+      allocWrapper(next.first, next.second);
+  }
+
+  void calculateInitialSubcircuit() {
+    auto control = start->getOperand(0);
+    auto target = start->getOperand(1);
+
+    addAnchorPoint(control, start);
+    addAnchorPoint(target, start);
+    while (!anchor_points.empty())
+      processNextAnchorPoint();
+  }
+
+  void pruneSubcircuit() {
+    for (auto *netlist : qubits)
+      if (netlist)
+        netlist->prune();
+    // Clean up
+    for (size_t i = 0; i < qubits.size(); i++) {
+      if (qubits[i] && !qubits[i]->hasOps()) {
+        delete qubits[i];
+        qubits[i] = nullptr;
+      }
+    }
+  }
+
+public:
+  /// @brief Constructs a subcircuit containing only RZ, NOT, CNOT, and Swap
+  /// gates, using the Netlist representation `netlist`.
+  /// @details A subcircuit is an connected portion of the netlist containing
+  /// only RZ, NOT, CNOT, and Swap gates.
+  ///
+  /// First, we construct an initial subcircuit:
+  /// We start by walking forward and backward along the netlist from the
+  /// initial anchor point, which is the control of `cnot`, and add the
```

Comment:
Just checking if the comment here is correct?

```suggestion
  /// initial anchor point, which is a `cnot`, and add the
```

Reply:
An anchor point is a particular point on the netlist, along some qubit at an operation. I think it is more correct to say "at `cnot`", but also to clarify that it is for a particular qubit. I'll work on clarifying the explanation here.